layer2-state-provider/history_commitment_provider.go

```diff
@@ -146,9 +177,167 @@ func (p *HistoryCommitmentProvider) HistoryCommitment(
+			ctx,
+			wasmModuleRoot,
+			Batch(assertionInboxMaxCount),
+			[]Height{Height(parentAssertionAfterStateBatch)},
+			option.Some[Height](Height(parentAssertionAfterStateBatch+commit.Height)))
+		if err != nil {
+			return false, err
+		}
+	default:
+		localCommit, err = p.HistoryCommitment(
+			ctx,
+			wasmModuleRoot,
+			Batch(assertionInboxMaxCount),
+			append(startHeights, 0),
+			option.Some[Height](Height(commit.Height)))
+		if err != nil {
+			return false, err
+		}
+	}
+	return localCommit.Height == commit.Height && localCommit.Merkle == commit.MerkleRoot, nil
+}
+
+var (
+	b32Arr, _ = abi.NewType("bytes32[]", "", nil)
+	// ProofArgs for submission to the protocol.
+	ProofArgs = abi.Arguments{
+		{Type: b32Arr, Name: "prefixExpansion"},
+		{Type: b32Arr, Name: "prefixProof"},
+	}
+)
+
+func (p *HistoryCommitmentProvider) PrefixProof(
+	ctx context.Context,
+	wasmModuleRoot common.Hash,
+	batch Batch,
+	startHeights []Height,
+	fromMessageNumber Height,
+	upToHeight option.Option[Height],
+) ([]byte, error) {
+	prefixLeaves, err := p.historyCommitmentImpl(
+		ctx,
+		wasmModuleRoot,
+		batch,
+		startHeights,
+		upToHeight)
+	if err != nil {
+		return nil, err
+	}
+	loSize := uint64(fromMessageNumber + 1)
+	hiSize := uint64(upToHeight.Unwrap() + 1)
+	if len(startHeights) == 1 {
+		loSize -= uint64(startHeights[0])
+		hiSize -= uint64(startHeights[0])
+	}
+	prefixExpansion, err := prefixproofs.ExpansionFromLeaves(prefixLeaves[:loSize])
+	if err != nil {
+		return nil, err
+	}
+	prefixProof, err := prefixproofs.GeneratePrefixProof(
+		loSize,
+		prefixExpansion,
+		prefixLeaves[loSize:hiSize],
+		prefixproofs.RootFetcherFromExpansion,
+	)
+	if err != nil {
+		return nil, err
+	}
+	_, numRead := prefixproofs.MerkleExpansionFromCompact(prefixProof, loSize)
+	onlyProof := prefixProof[numRead:]
+	return ProofArgs.Pack(&prefixExpansion, &onlyProof)
+}
+
+func (p *HistoryCommitmentProvider) OneStepProofData(
+	ctx context.Context,
+	wasmModuleRoot common.Hash,
+	startHeights []Height,
+	upToHeight Height,
+) (*protocol.OneStepData, []common.Hash, []common.Hash, error) {
+	endCommit, err := p.HistoryCommitment(
+		ctx,
+		wasmModuleRoot,
+		0,
+		startHeights,
+		option.Some(upToHeight+1),
+	)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	startCommit, err := p.HistoryCommitment(
+		ctx,
+		wasmModuleRoot,
+		0,
+		startHeights,
+		option.Some(upToHeight),
+	)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Compute the exact start point of where we need to execute
+	// the machine from the inputs, and figure out, in what increments, we need to do so.
+	machineIndex, err := p.computeMachineStartIndex(startHeights)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	machineIndex += OpcodeIndex(upToHeight)
+	hashes, err := p.machineHashCollector.CollectMachineHashes(
+		ctx,
+		&HashCollectorConfig{
+			WasmModuleRoot:    wasmModuleRoot,
+			MessageNumber:     startHeights[0],
+			StepHeights:       startHeights[1:],
```

Comment:
do we need startHeights length check?

Reply:
Fixed the API so that we no longer need to do this janky stuff :)